src/plugin.ts

```diff
@@ -372,33 +378,80 @@ function runtimeErrorIIFE(babel: typeof Babel, replacements: { ERROR_MESSAGE: st
 
 function buildScopeLocals(
   userTypedOptions: Record<string, unknown>,
-  formatOptions: ModuleConfig,
-  templateContent: string
+  formatOptions: ModuleConfig
 ): ScopeLocals {
   if (formatOptions.rfc931Support && userTypedOptions.eval) {
-    return discoverLocals(templateContent);
+    return new ScopeLocals();
   } else if (userTypedOptions.scope) {
     return userTypedOptions.scope as ScopeLocals;
   } else {
     return new ScopeLocals();
   }
 }
 
-function discoverLocals(templateContent: string): ScopeLocals {
-  // this is wrong, but the right thing is unreleased in
-  // https://github.com/glimmerjs/glimmer-vm/pull/1421, so for the moment I'm
-  // sticking with the exact behavior that ember-templates-imports has.
-  //
-  // (the reason it's wrong is that the correct answer depends on not just the
-  // template, but the ambient javascript scope. Anything in locals needs to win
-  // over ember keywords. Otherwise we can never introduce new keywords.)
-  let scopeLocals = new ScopeLocals();
-  for (let local of getTemplateLocals(templateContent)) {
-    if (local.match(/^[$A-Z_][0-9A-Z_$]*$/i)) {
-      scopeLocals.add(local);
+function discoverLocals<EnvSpecificOptions>(
+  jsPath: NodePath<t.Expression>,
+  state: State<EnvSpecificOptions>,
+  locals: string[],
+  scope: ScopeLocals
+) {
+  function add(name: string) {
+    if (locals.includes(name)) return;
+    locals.push(name);
+  }
+
+  let astScope: string[] = [];
+  function isInScope(name: string) {
+    return astScope.flat().includes(name);
+  }
+
+  function isInJsScope(name: string) {
+    if (jsPath.scope.getBinding(name)) return true;
+    if (['this', 'globalThis'].includes(name)) return true;
+    if (scope.locals.includes(name)) return true;
+    if (state.originalImportedNames.has(name)) {
+      return true;
     }
+    return false;
   }
-  return scopeLocals;
+
+  return (_env: ASTPluginEnvironment): { name: string; visitor: NodeVisitor } => {
+    return {
+      name: 'discover-locals',
+      visitor: {
+        All: {
+          enter(_node, path) {
+            const blockParams = (path.parentNode as any)?.blockParams;
+            if (blockParams && ['children', 'body'].includes(path.parentKey!)) {
+              astScope.push(blockParams);
```

Comment:
we just push the block params array, since its the same instance we can easily remove it at exit